typed-transitions/test/Test/Protocol/Codec/Coherent.hs

```diff
@@ -0,0 +1,440 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE StandaloneDeriving #-}
+
+{-# OPTIONS_GHC "-fwarn-incomplete-patterns" #-}
+
+module Test.Protocol.Codec.Coherent where
+
+import Data.Either (partitionEithers)
+import Data.List (intercalate, sortBy)
+import Data.List.NonEmpty (NonEmpty)
+import qualified Data.List.NonEmpty as NE
+import Data.Functor.Identity (Identity (..))
+import Data.Kind (Type)
+import Data.Maybe (mapMaybe)
+import Data.Type.Equality
+
+import Test.QuickCheck
+
+import Protocol.Codec
+import Protocol.Path
+
+-- | Test the coherence of a 'CodecTree' (use 'codecTreePaths' to get one).
+--
+-- This test checks 2 properties:
+-- 
+-- 1. Local coherence of each encoder/decoder. If the decoder fails to invert
+--    the encoder, then it's locally incoherent.
+-- 2. Global coherence of the codec itself. If any 2 paths disagree on the
+--    encoding at the same spot in the path, then the codec is globally
+--    incoherent.
```

Comment:
Isn't 1. a consequence of 2.?

Reply:
I don't think so. 2 is really about the encoders, and that the decoder paths sort of follow the encoder paths and don't change the encoders themselves.

Imagine the state machine with 1 state, 1 transition. There's a codec which always encodes that transition to one particular representation, say `"0"`. But we're free to make 2 different codes by choosing different decoders. Clearly, there's only one sensible decoder: the one that inverts this encoder. But one could define and use one that doesn't, in which case we have 2 but not 1.